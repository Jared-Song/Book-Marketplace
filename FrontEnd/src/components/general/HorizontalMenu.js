import { Button, Grid } from "@material-ui/core";
import React from "react";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles((theme) => ({
  selectedItem: {
    width: "100%",
    backgroundColor: "#EBF0FF",
    borderRadius: 15,
    cursor: 'pointer'
  },
  item: {
    width: "100%",
    cursor: 'pointer'
  },
}));

// menuItem shoud looks like: [{title: String, selected: Boolean!, action: () => void}]
export default function HorizontalMenu({ menuItems }) {
  const classes = useStyles();

  return (
    <Grid
      container
      direction="column"
      justifyContent="flex-start"
      alignItems="flex-start"
      spacing={2}
    >
      {menuItems.map((menuItem) => (
        <Grid
          item
          xs={12}
          key={menuItem.title}
          className={menuItem.selected ? classes.selectedItem : classes.item}
        >
          <Button disabled={menuItem.selected} onClick={menuItem.onClick}>{menuItem.title}</Button>
        </Grid>
      ))}
    </Grid>
  );
}
