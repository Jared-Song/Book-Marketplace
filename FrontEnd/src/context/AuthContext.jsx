import React from "react";
import { useJwt } from "react-jwt";

const AuthContext = React.createContext(null);

export const AuthProvider = ({ token, children, loading }) => {
  const [currentToken, setToken] = React.useState(token);
  const { decodedToken } = useJwt(currentToken);
  const [isLoading, setIsLoading] = React.useState(loading);
  React.useEffect(() => {
    setToken(token);
  }, [token]);
  return (
    <AuthContext.Provider
      value={{ currentUser: decodedToken, setToken, isLoading }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export const useCurrentUser = () => React.useContext(AuthContext);
