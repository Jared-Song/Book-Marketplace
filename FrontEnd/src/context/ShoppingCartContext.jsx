import _, { set } from "lodash";
import React from "react";

const ShoppingCartContext = React.createContext(null);

export const ShoppingCartProvider = ({ children }) => {
  const storageItem = typeof window !== "undefined" ? localStorage.getItem("cart") : undefined;
  const [items, setItems] = React.useState(storageItem ? JSON.parse(storageItem) : []);
  React.useEffect(() => {
    localStorage.setItem("cart", JSON.stringify(items));
  }, [items]);
  const removeAllItems = () => {
    setItems([])
    localStorage.clear()
  }

  const removeShoppingCartItem = (id) => {
    const updatedItems = _.remove(items, (item) => {
      return item.id !== id
    })
    setItems(updatedItems)
  }
  const editShoppingCartItem = (newItem) => {
    const updatedItems = items.map((item) => {
      if (newItem.id === item.id) {
        return newItem; 
      }
      return item
    })
    setItems(updatedItems);
  }
  const addIntoShoppingCart = (newItem) => {
    const itemExist = items.filter((item) => item.id === newItem.id);
    if (itemExist.length > 0) {
      const updatedItems = items.map((item) => {
        if (item.id === newItem.id) {
          return {
            ...item,
            quantity: item.quantity + parseInt(newItem.quantity)
          }
        }
        return item;
      })
      setItems(updatedItems);
    } else {
      setItems([
        ...items,
        {
          ...newItem,
          quantity: parseInt(newItem.quantity),
        },
      ]);
    }
  }
  return (
    <ShoppingCartContext.Provider
      value={{
        cartItems: items,
        addIntoShoppingCart,
        editShoppingCartItem,
        removeShoppingCartItem,
        removeAllItems,
      }}
    >
      {children}
    </ShoppingCartContext.Provider>
  );
}

export const useShoppingCart = () => React.useContext(ShoppingCartContext);