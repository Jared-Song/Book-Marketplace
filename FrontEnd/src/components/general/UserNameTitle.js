import Typography from "@material-ui/core/Typography";
import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import FaceIcon from "@material-ui/icons/Face";
import { useCurrentUser } from "../../context/AuthContext";
import { useRouter } from "next/router";

const useStyles = makeStyles((theme) => ({
  title: {
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center",
  },
}));
export default function UserNameTitle() {
  const router = useRouter();

  const classes = useStyles();
  const { currentUser } = useCurrentUser();

  if (currentUser) {
    return (
      <div className={classes.title}>
        <FaceIcon style={{ fontSize: 50 }} />
        <Typography variant="h4">Hello, {currentUser.username}!</Typography>
      </div>
    );
  }
  return null;
}
