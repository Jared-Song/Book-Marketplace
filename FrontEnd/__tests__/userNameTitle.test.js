import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import HorizontalMenu from "../src/components/general/HorizontalMenu";
import PrimarySearchAppBar from "../src/components/general/MenuBar";
import { AuthProvider } from "../src/context/AuthContext";
import { useRouter } from "next/router";
import UserNameTitle from "../src/components/general/UserNameTitle";

it("renders UserNameTitle when there is no user", () => {
  const tree = renderer
    .create(
      <AuthProvider token="">
        <UserNameTitle />
      </AuthProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
