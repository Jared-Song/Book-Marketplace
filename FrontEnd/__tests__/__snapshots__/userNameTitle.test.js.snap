// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders UserNameTitle when there is no user 1`] = `null`;
